### PR TITLE
mypy==1.13.0

### PR DESCRIPTION
**Pull request by JakobSpahn**
Depends on https://github.com/tinygrad/tinygrad/pull/7863, https://github.com/tinygrad/tinygrad/pull/7866.
Finalizes the attempt of https://github.com/tinygrad/tinygrad/pull/7846.

### CODE REVIEW

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -281,7 +281,7 @@ def data(self) -> memoryview:
     assert self.dtype.fmt is not None, f"no fmt dtype for {self.dtype}"
     assert all_int(self.shape), f"no data if shape is symbolic, {self.shape=}"
     if TYPE_CHECKING or sys.version_info < (3, 12): assert self.dtype.fmt != "e"
-    return self._data().cast(self.dtype.fmt) if 0 in self.shape else self._data().cast(self.dtype.fmt, self.shape)
+    return cast(memoryview, self._data().cast(self.dtype.fmt) if 0 in self.shape else self._data().cast(self.dtype.fmt, self.shape))
```
Comment (JakobSpahn) [inline tinygrad/tensor.py:284]: Cast `memoryview[int] | memoryview[float] | memoryview[bool]` to `memoryview`.

Comment (geohot) [reply tinygrad/tensor.py:284]: Any reason to not make return type `memoryview[int] | memoryview[float] | memoryview[bool]`?

Comment (JakobSpahn) [reply tinygrad/tensor.py:284]: In the [typeshed stub](https://github.com/python/mypy/blob/v1.13.0/mypy/typeshed/stdlib/builtins.pyi#L739), `memoryview` is generically defined as:

```python
_I = TypeVar("_I", default=int)  
class memoryview(Sequence[_I]):
  ...

```
Which leads to the typechecker substituting every `memoryview` type decl with `memoryview[int]`.

Changing the return type will require to change the types of a lot of variables/function args from `memoryview` to the broader `memoryview[int] | memoryview[float] | memoryview[bool]`. For example all the `Allocator._copyin` implementations (and so on).

I can use a type alias such as `MemoryviewType = memoryview[int] | memoryview[float] | memoryview[bool]` but that will be very confusing with both `MemoryviewType` and `memoryview` scattered across the codebase.

Imo the cleanest solution is to return a `memoryview` or a `memoryview[Any]`

Comment (geohot) [reply tinygrad/tensor.py:284]: kk, cast is fine

**Comment by geohot** [comment]
Is there a reason this is draft?

File: tinygrad/runtime/graph/hcq.py [hunk 2]
```diff
@@ -54,6 +54,8 @@ def __init__(self, jit_cache: List[ExecItem], input_rawbuffers: List[Buffer], va
 
     for j,ji in enumerate(jit_cache):
       enqueue_dev = ji.prg.dev if (is_exec_prg:=isinstance(ji.prg, CompiledRunner)) else Device[ji.bufs[1].device] #type:ignore
+      assert isinstance(enqueue_dev, HCQCompiled)
+      assert (enqueue_dev.hw_copy_queue_t is not None) and issubclass(enqueue_dev.hw_compute_queue_t, HWQueue)
```
[the remark this answers is not in the document]
Comment (geohot) [reply tinygrad/runtime/graph/hcq.py:58]: Still should be one line

**Comment by JakobSpahn** [comment]
Changed to one line assert, should be ready now

File: tinygrad/runtime/graph/hcq.py [hunk 3]
```diff
@@ -54,6 +54,7 @@ def __init__(self, jit_cache: List[ExecItem], input_rawbuffers: List[Buffer], va
 
     for j,ji in enumerate(jit_cache):
       enqueue_dev = ji.prg.dev if (is_exec_prg:=isinstance(ji.prg, CompiledRunner)) else Device[ji.bufs[1].device] #type:ignore
+      assert isinstance(enqueue_dev, HCQCompiled) and issubclass(enqueue_dev.hw_compute_queue_t, HWQueue) and enqueue_dev.hw_copy_queue_t is not None
```
Comment (geohot) [inline tinygrad/runtime/graph/hcq.py:57]: Err, how sure are you that this is right? What if it doesn't have a copy queue?

File: tinygrad/runtime/graph/hcq.py [hunk 4]
```diff
@@ -54,7 +54,11 @@ def __init__(self, jit_cache: List[ExecItem], input_rawbuffers: List[Buffer], va
 
     for j,ji in enumerate(jit_cache):
       enqueue_dev = ji.prg.dev if (is_exec_prg:=isinstance(ji.prg, CompiledRunner)) else Device[ji.bufs[1].device] #type:ignore
```
Comment (JakobSpahn) [inline tinygrad/runtime/graph/hcq.py:56]: Kinda hacky, but could save one line by removing the following assert and declaring `enqueue_dev: HCQCompiled`.

File: tinygrad/runtime/graph/hcq.py (hunk 3 above)
Comment (JakobSpahn) [reply tinygrad/runtime/graph/hcq.py]: Sorry I missed that QCOM had a graph

File: tinygrad/runtime/graph/hcq.py [hunk 5]
```diff
@@ -54,7 +54,11 @@ def __init__(self, jit_cache: List[ExecItem], input_rawbuffers: List[Buffer], va
...
-      enqueue_queue = self.comp_queues[enqueue_dev] if is_exec_prg else self.copy_queues.setdefault(enqueue_dev, enqueue_dev.hw_copy_queue_t())
+      assert isinstance(enqueue_dev, HCQCompiled)
+      if is_exec_prg: enqueue_queue = self.comp_queues[enqueue_dev]
+      else:
+        assert issubclass(enqueue_dev.hw_compute_queue_t, HWQueue) and enqueue_dev.hw_copy_queue_t is not None
+        enqueue_queue = self.copy_queues.setdefault(enqueue_dev, enqueue_dev.hw_copy_queue_t())
```
Comment (geohot) [inline tinygrad/runtime/graph/hcq.py:61]: This is unreadable and needs refactoring

Comment (JakobSpahn) [reply tinygrad/runtime/graph/hcq.py:61]: Going to open up a new PR with the other changes and convert this one to draft.

Comment (JakobSpahn) [reply tinygrad/runtime/graph/hcq.py:62]: Let me know if it is readable enough now.
Unfortunately, mypy doesn't have the path sensitivity for something with as few lines as possible like:
```python
if not is_exec_prg: assert (enqueue_dev.hw_copy_queue_t is not None)
enqueue_queue = self.comp_queues[enqueue_dev] if is_exec_prg else self.copy_queues.setdefault(enqueue_dev, enqueue_dev.hw_copy_queue_t())
```

**Comment by geohot** [comment]
Assuming benchmark tests pass, the bounty is yours! E-mail george@tinygrad.org to claim, can pay via PayPal or USDC on ETH

**Pull request merged**